### PR TITLE
gromacs: oneapi does not always require gcc

### PR DESCRIPTION
**Pull request by mabraham**

### CODE REVIEW

**Comment by mabraham** [comment]
@junghans I don't know the history here, but certainly oneapi does not ship a bundled gcc (or runtime) for the last few years, so I think this gcc dependency is much too heavy. In the meantime, spack has evolved both `intel-oneapi-runtime` and `gcc-runtime`, and with the code of this MR I saw:
```
$ spack spec gromacs@git.intel-main=2025 %oneapi
 -   gromacs@git.intel-main=2025%oneapi@2025.0.0~cp2k~cuda~cycle_subcounters~double+gmxapi+hwloc~intel_provided_gcc~ipo~mdrun_only+mpi+nblib~nosuffix~opencl+openmp~relaxed_double_precision+shared build_system=cmake build_type=Release generator=make openmp_max_threads=none arch=linux-sles15-icelake
[e]      ^cmake@3.28.1%oneapi@2025.0.0~doc+ncurses+ownlibs~qtgui build_system=generic build_type=Release patches=dbc3892 arch=linux-sles15-icelake
[e]      ^glibc@2.31%gcc@12.2.1 build_system=autotools arch=linux-sles15-icelake
[+]      ^gmake@4.4.1%gcc@12.2.1~guile build_system=generic arch=linux-sles15-icelake
[+]          ^gcc-runtime@12.2.1%gcc@12.2.1 build_system=generic arch=linux-sles15-icelake
[e]          ^glibc@2.31%oneapi@2025.0.0 build_system=autotools arch=linux-sles15-icelake
[+]      ^hwloc@2.11.1%gcc@12.2.1~cairo~cuda~gl~libudev~libxml2~nvml+oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-sles15-icelake
[+]          ^libpciaccess@0.17%gcc@12.2.1 build_system=autotools arch=linux-sles15-icelake
[+]              ^util-macros@1.20.1%gcc@12.2.1 build_system=autotools arch=linux-sles15-icelake
[+]          ^ncurses@6.5%gcc@12.2.1~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-sles15-icelake
[+]          ^oneapi-level-zero@1.17.2%gcc@12.2.1~ipo build_system=cmake build_type=Release generator=make arch=linux-sles15-icelake
[e]              ^cmake@3.28.1%gcc@12.2.1~doc+ncurses+ownlibs~qtgui build_system=generic build_type=Release patches=dbc3892 arch=linux-sles15-icelake
[e]      ^intel-oneapi-mkl@2025.0%oneapi@2025.0.0~cluster+envmods~gfortran~ilp64+shared build_system=generic mpi_family=none threads=none arch=linux-sles15-icelake
[e]      ^intel-oneapi-mpi@2021.14%oneapi@2025.0.0~classic-names+envmods~external-libfabric~generic-names~ilp64 build_system=generic arch=linux-sles15-icelake
 -       ^intel-oneapi-runtime@2025.0.0%oneapi@2025.0.0 build_system=generic arch=linux-sles15-icelake
 -           ^gcc-runtime@12.2.1%gcc@12.2.1 build_system=generic arch=linux-sles15-icelake
[+]      ^pkgconf@2.2.0%gcc@12.2.1 build_system=autotools arch=linux-sles15-icelake
```
That looks good to me. Thoughts?

I have not tried to test building GROMACS with a oneapi installed via spack, but I believe it works the same.

**Review by junghans — DISMISSED** [review]

**Comment by junghans** [comment]
Makes sense!

**Comment by junghans** [comment]
@mabraham 
```
RuntimeError: failed to concretize `gromacs%oneapi arch=None-None-x86_64_v3` for the following reasons:
     1. gromacs: Only attempt to find gcc libs for Intel compiler if Intel compiler is used.
```

**Comment by mabraham** [comment]
> @mabraham
> 
> ```
> RuntimeError: failed to concretize `gromacs%oneapi arch=None-None-x86_64_v3` for the following reasons:
>      1. gromacs: Only attempt to find gcc libs for Intel compiler if Intel compiler is used.
> ```

I don't think that error is possible from that spec and this code, because it is only issued with the new code when `%intel` classic compile is used. Can you check what spack commit you were using? *edit* I see, that was in CI. Weird

**Comment by mabraham** [comment]
> > @mabraham
> > ```
> > RuntimeError: failed to concretize `gromacs%oneapi arch=None-None-x86_64_v3` for the following reasons:
> >      1. gromacs: Only attempt to find gcc libs for Intel compiler if Intel compiler is used.
> > ```
> 
> I don't think that error is possible from that spec and this code, because it is only issued with the new code when `%intel` classic compile is used. Can you check what spack commit you were using? _edit_ I see, that was in CI. Weird

I still think the only way we can get that error is if they had a stale GROMACS `package.py` or some other bug (note that the e4s oneapi build went fine). But since I anyway planned to remove this error in follow-up work, I'll do that now and see what we see in CI.

**Comment by mabraham** [comment]
> I have not tried to test building GROMACS with a oneapi installed via spack, but I believe it works the same.

I have now tried that and it also does not require gcc, and works fine with this MR

**Comment by mabraham** [comment]
Now we see
```
RuntimeError: failed to concretize `gromacs%oneapi arch=None-None-x86_64_v4` for the following reasons:
     1. Cannot set the required compiler: gromacs%oneapi
```
from https://gitlab.spack.io/spack/spack/-/jobs/13762050. That continues to suggest something is broken, e.g. no `%oneapi` is present in the CI build image.

That speculation seems consistent with https://gitlab.spack.io/spack/spack/-/jobs/13762050#L35

I see that https://github.com/spack/gitlab-runners/pull/52 recently rebuilt these images, and perhaps this is the first PR that finds a problem with that. @alalazo I have made a suggestion at https://github.com/spack/gitlab-runners/pull/52#discussion_r1850112406. Does it makes sense to try that, please?

**Comment by stephenmsachs** [comment]
Let me sched some light on why this was [added in the first place](https://github.com/spack/spack/commit/d91ec8500f7f8265c023b217bce0c0ca34a81a24). If an Intel compiler (classic or oneapi) is used, the `cmake` will try to infer where it can find `g++` as well as the standard libraries by looking at the `--gcc-toolchain` and `--gcc-name` flags given as Cmake variables.

If the Intel classic or oneapi compiler was installed via spack, then those flags are hidden in `icp{c,x}.cfg`. This means Cmake potentially does not find the correct `g++` and uses the wrong libraries. The `+intel_provided_gcc` flag enables using a Spack installed Intel classic (and oenapi) compiler on top of a non-standard (i.e. not `/usr/bin/g++`) `g++`.

> Now we see
> 
> ```
> RuntimeError: failed to concretize `gromacs%oneapi arch=None-None-x86_64_v4` for the following reasons:
>      1. Cannot set the required compiler: gromacs%oneapi
> ```

I believe that this error comes from the fact that the [new restriction to only intel](https://github.com/mabraham/spack/blob/ed71ca1ee55e87daca9a1ea006386e1e6fd6e33d/var/spack/repos/builtin/packages/gromacs/package.py#L294) prevents the concretizer to [use oneapi in conjunction with +intel_provided_gcc](https://github.com/spack/spack/blob/0ff3e863151a2c74ac202eefc0b4d3f65b39a85e/share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-x86_64_v4/packages.yaml#L24).

> 
> from https://gitlab.spack.io/spack/spack/-/jobs/13762050. That continues to suggest something is broken, e.g. no `%oneapi` is present in the CI build image.
> 
> That speculation seems consistent with https://gitlab.spack.io/spack/spack/-/jobs/13762050#L35

Here only the unknown (system) compilers are detected. Intel and oneapi compilers are already known form the container build time and their configuration is simply [copied](https://github.com/spack/spack/blob/0ff3e863151a2c74ac202eefc0b4d3f65b39a85e/share/spack/gitlab/cloud_pipelines/scripts/pcluster/setup-pcluster.sh#L77) to save time.

> 
> I see that [spack/gitlab-runners#52](https://github.com/spack/gitlab-runners/pull/52) recently rebuilt these images, and perhaps this is the first PR that finds a problem with that. @alalazo I have made a suggestion at [spack/gitlab-runners#52 (comment)](https://github.com/spack/gitlab-runners/pull/52#discussion_r1850112406). Does it makes sense to try that, please?

No, this is not the first time we using these compilers. The same pipeline has been building `lammps`, `palace`, and other codes with `oneapi`.

**Comment by stephenmsachs** [comment]
The title of this PR "gromacs: oneapi does not always require gcc" is kind of misleading. The `+intel_provided_gcc`variant is disabled by default, so there should be nothing stopping you from building gromacs with oneapi *without* gcc.

**Comment by mabraham** [comment]
> The title of this PR "gromacs: oneapi does not always require gcc" is kind of misleading. The `+intel_provided_gcc`variant is disabled by default, so there should be nothing stopping you from building gromacs with oneapi _without_ gcc.

Thanks for the feedback! The former code had `depends_on("gcc", when="%oneapi ~intel_provided_gcc")` which I observed to lead to a `gromacs %oneapi` spec requiring a `gcc` package. That's the major issue I'm seeking to resolve.

**Comment by mabraham** [comment]
> Let me sched some light on why this was [added in the first place](https://github.com/spack/spack/commit/d91ec8500f7f8265c023b217bce0c0ca34a81a24). If an Intel compiler (classic or oneapi) is used, the `cmake` will try to infer where it can find `g++` as well as the standard libraries by looking at the `--gcc-toolchain` and `--gcc-name` flags given as Cmake variables.

Thank you for shedding light here. Please also note that (at least these days) GROMACS also does not need to do anything regarding `libstdc++` if the system and compiler is already set up to use a sufficiently modern one (see https://github.com/gromacs/gromacs/blob/cb6b311c39fc726a72170c4593b754c8d0a492ac/cmake/FindLibStdCpp.cmake#L78). I am using an OpenSuSe cluster where the system `gcc` is version `7.x` and would definitely fail this test, but there is a more recent `libstdc++` installed, which passes the test. I think it is useful to users of spack+gromacs that in such cases (and also simple ones where the default system `libstdc++` Just Works) that `gromacs %oneapi` also just works without requiring either a `gcc` dependency or to know to manipulate `intel_provided_gcc` variant.

I think it's also important that we keep spack+gromacs working on clusters where the `libstdc++` is not new enough for GROMACS, for two reasons:
* that old specs can keep working, and
* that compilers other than `gcc` may have fresh need of such support, because GROMACS will shortly require C++20

**Comment by mabraham** [comment]
I think `intel-oneapi-runtime` and `gcc-runtime` do a good job of modelling how a package like `gromacs` should be able to resolve the dependency on `libstdc++`. I think the GROMACS `package.py` should be able to declare requirements on libstdc++ (which should be versioned as GROMACS requirements evolve) via `gcc-runtime`, and only if no currently available `gcc-runtime` can do the job is one obtained via requiring a new installation of `gcc` package. Either way, GROMACS can be configured to use the matching `libstdc++.so`, *and* the user can have control via `^gcc-runtime@x.y` if they need that. I'll see if I can make that work. Any feedback?

**Comment by stephenmsachs** [comment]
> I think `intel-oneapi-runtime` and `gcc-runtime` do a good job of modelling how a package like `gromacs` should be able to resolve the dependency on `libstdc++`. I think the GROMACS `package.py` should be able to declare requirements on libstdc++ (which should be versioned as GROMACS requirements evolve) via `gcc-runtime`, and only if no currently available `gcc-runtime` can do the job is one obtained via requiring a new installation of `gcc` package. Either way, GROMACS can be configured to use the matching `libstdc++.so`, _and_ the user can have control via `^gcc-runtime@x.y` if they need that. I'll see if I can make that work. Any feedback?

Yes, I think using an appropriate `gcc-runtime` would be the best solution. The `intel_provided_gcc` variant predates the run times, so this was not possible at that moment. So I guess we could add something like 
```
options.append("-DCMAKE_CXX_FLAGS=--gcc-toolchain {}".format(self.spec["gcc-runtime"].prefix))
```


>Thanks for the feedback! The former code had depends_on("gcc", when="%oneapi ~intel_provided_gcc") which I observed to lead to a gromacs %oneapi spec requiring a gcc package. That's the major issue I'm seeking to resolve.

I believe that is an artifact of trying to satisfy [Cmake looking for an explicit g++](https://github.com/gromacs/gromacs/blame/cb6b311c39fc726a72170c4593b754c8d0a492ac/cmake/FindLibStdCpp.cmake#L109). This happens in case `--gcc-toolchain` is not found in the `CMAKE_CXX_FLAGS` flags (e.g. Spack installed `oneapi` compiler). `find_program()` will only search default search paths, so this installation will fail on systems without something like `/usr/bin/g++`.

**Comment by stephenmsachs** [comment]
I agree that if this [check_cxx_source_compiles](https://github.com/gromacs/gromacs/blame/cb6b311c39fc726a72170c4593b754c8d0a492ac/cmake/FindLibStdCpp.cmake#L76C10-L76C10) function returns true we don't have to worry about the `g++` location anymore and we don't need ` depends_on("gcc", when="%oneapi ~intel_provided_gcc")` or ` depends_on("gcc", when="%intel ~intel_provided_gcc")`. But I don't know how to make that destinction in the Spack package.

**Comment by mabraham** [comment]
I've made a partial attempt at depending on a suitable `gcc-runtime`. I do not yet have a way of deciding that because there is no existing suitable `gcc-runtime` that the spec should imply installing `gcc` package.

File: var/spack/repos/builtin/packages/gromacs/package.py [hunk 1]
```diff
@@ -540,7 +546,7 @@ def cmake_args(self):
             ):
                 with open(".".join([os.environ["SPACK_CXX"], "cfg"]), "r") as f:
                     options.append("-DCMAKE_CXX_FLAGS={}".format(f.read()))
-            else:
+            elif self.spec.satisfies("+gcc"):
```
Comment (stephenmsachs) [inline var/spack/repos/builtin/packages/gromacs/package.py]: Do you mean `^gcc`? There is no `gcc` variant.

**Comment by stephenmsachs** [comment]
> I've made a partial attempt at depending on a suitable gcc-runtime. I do not yet have a way of deciding that because there is no existing suitable gcc-runtime that the spec should imply installing gcc package.

Looks good to me.

File: var/spack/repos/builtin/packages/gromacs/package.py (hunk 1 above)
Comment (junghans) [inline var/spack/repos/builtin/packages/gromacs/package.py]:
```suggestion
            elif self.spec.satisfies("^gcc"):
```

File: var/spack/repos/builtin/packages/gromacs/package.py [hunk 2]
```diff
@@ -310,8 +310,14 @@ class Gromacs(CMakePackage, CudaPackage):
     depends_on("sycl", when="+sycl")
     depends_on("lapack")
     depends_on("blas")
-    depends_on("gcc", when="%oneapi ~intel_provided_gcc")
     depends_on("gcc", when="%intel ~intel_provided_gcc")
+    # TODO this can be expanded to all clang-based compilers once
```
Comment (junghans) [inline var/spack/repos/builtin/packages/gromacs/package.py:314]: @mabraham do you want to add more things here?

Comment (mabraham) [reply var/spack/repos/builtin/packages/gromacs/package.py:314]: In a later MR, I think

**Comment by mabraham** [comment]
> Yes, I think using an appropriate `gcc-runtime` would be the best solution. The `intel_provided_gcc` variant predates the run times, so this was not possible at that moment. So I guess we could add something like
> 
> ```
> options.append("-DCMAKE_CXX_FLAGS=--gcc-toolchain {}".format(self.spec["gcc-runtime"].prefix))
> ```

I tried this, but it did not work. GROMACS CMake call could not identify the compiler because the test program failed to find `libcrtS.so`. So I deduce that `--gcc-toolchain` really does need to be able to find more than the handful of libraries currently in `self.spec["gcc-runtime"].prefix.lib`. I imagine a future world where `gcc-runtime` has an optional pointer to the related installation of g++ that could be used here, but we don't currently have that. If we find we need it, then we should probably ask for that feature added to Spack.

File: var/spack/repos/builtin/packages/gromacs/package.py [hunk 3]
```diff
@@ -540,7 +546,7 @@ def cmake_args(self):
             ):
                 with open(".".join([os.environ["SPACK_CXX"], "cfg"]), "r") as f:
                     options.append("-DCMAKE_CXX_FLAGS={}".format(f.read()))
-            else:
+            elif self.spec.satisfies("^gcc"):
```
Comment (junghans) [inline var/spack/repos/builtin/packages/gromacs/package.py:549]:
```suggestion
            elif self.spec.satisfies("%gcc"):
```
I think I got this wrong!

Comment (mabraham) [reply var/spack/repos/builtin/packages/gromacs/package.py:549]: No, the relevant case is where we *compiled* with `%oneapi` and compiled a gcc (somehow) to get a satisfactory `gcc-runtime` from it. So it's definitely `+gcc` or `^gcc`

Comment (mabraham) [reply var/spack/repos/builtin/packages/gromacs/package.py:549]: I'd like to write the package to say "if you haven't got a satisfactory `gcc-runtime` right now, go compile a `gcc`" but I don't think there is a way to do it.

**Review by junghans — APPROVED** [review]

**Comment by junghans** [comment]
@mabraham can we merge this?

**Comment by mabraham** [comment]
> @mabraham can we merge this?

yes!

**Pull request merged**